File: sim/NationRise.Core/Ai/WarPlanner.cs
using NationRise.Core.Diplomacy;
using NationRise.Core.Military;
using NationRise.Core.World;

namespace NationRise.Core.Ai;

public readonly record struct TargetChoice(int Province, float Value, string Reason);

/*
   Decides where an army should go once war is under way. Deciding to fight and
   deciding where to fight are separate problems: the first is diplomacy, this
   is operations, and conflating them produces armies that march at whatever
   enemy province happens to be first in the array.
*/
public sealed class WarPlanner(WorldState world, Relations relations, Data.ProvinceGraph land)
{
    public const float CityWeight = 40f;
    public const float UndefendedWeight = 25f;
    public const float DistancePenalty = 3f;
    public const float FrontlineWeight = 15f;

    public TargetChoice? ChooseTarget(Army army, IReadOnlyDictionary<int, Army> allArmies)
    {
        ArgumentNullException.ThrowIfNull(army);

        var garrisons = new Dictionary<int, int>();
        foreach (Army other in allArmies.Values)
        {
            if (other.IsDestroyed || other.Nation == army.Nation)
            {
                continue;
            }

            garrisons[other.Province] = garrisons.GetValueOrDefault(other.Province) + other.Count;
        }

        TargetChoice? best = null;
        float bestValue = float.NegativeInfinity;

        for (int province = 0; province < world.Provinces.Count; province++)
        {
            ushort owner = world.Provinces.Controller[province];
            if (owner == ProvinceStore.NoOwner || !relations.AtWar(army.Nation, owner))
            {
                continue;
            }

            float value = ValueOf(province, army.Province, garrisons, out string reason);
            if (value > bestValue)
            {
                bestValue = value;
                best = new TargetChoice(province, value, reason);
            }
        }

        return best;
    }

    private float ValueOf(int province, int from, Dictionary<int, int> garrisons, out string reason)
    {
        float value = 10f;
        reason = "enemy ground";

        if (world.Provinces.IsCity[province])
        {
            value += CityWeight + world.Provinces.Population[province] * 4f;
            reason = "enemy city";
        }

        int defenders = garrisons.GetValueOrDefault(province);
        if (defenders == 0)
        {
            value += UndefendedWeight;
            if (reason == "enemy ground")
            {
                reason = "undefended";
            }
        }
        else
        {
            value -= defenders * 6f;
        }

        /* Provinces bordering ground we already hold are worth more than a
           prize deep inside enemy territory: a front that advances together
           can be supplied, a lone spearhead cannot. */
        if (BordersOwnTerritory(province, world.Provinces.Controller[from]))
        {
            value += FrontlineWeight;
        }

        value -= ApproximateDistance(from, province) * DistancePenalty;
        return value;
    }

    private bool BordersOwnTerritory(int province, ushort nation)
    {
        foreach (ushort neighbour in land.NeighboursOf(province))
        {
            if (world.Provinces.Controller[neighbour] == nation)
            {
                return true;
            }
        }

        return false;
    }

    /* Breadth-first hop count, capped: a full path search for every candidate
       province would cost more than the decision is worth. */
    private int ApproximateDistance(int from, int to)
    {
        const int Cap = 8;

        var seen = new HashSet<int> { from };
        var frontier = new Queue<(int Province, int Depth)>();
        frontier.Enqueue((from, 0));

        while (frontier.Count > 0)
        {
            (int province, int depth) = frontier.Dequeue();
            if (province == to)
            {
                return depth;
            }

            if (depth >= Cap)
            {
                continue;
            }

            foreach (ushort neighbour in land.NeighboursOf(province))
            {
                if (seen.Add(neighbour))
                {
                    frontier.Enqueue((neighbour, depth + 1));
                }
            }
        }

        return Cap;
    }
}

File: web/src/sim/ai/WarPlanner.ts
import { addF32, f32, mulF32, subF32 } from '~/sim/determinism/float32'
import type { ProvinceGraph } from '~/sim/data/WorldFile'
import type { Relations } from '~/sim/diplomacy/Relation'
import type { Army } from '~/sim/military/Army'
import { NO_OWNER } from '~/sim/world/ProvinceStore'
import type { WorldState } from '~/sim/world/WorldState'

export interface TargetChoice {
  readonly province: number
  readonly value: number
  readonly reason: string
}

export const CITY_WEIGHT = f32(40)
export const UNDEFENDED_WEIGHT = f32(25)
export const DISTANCE_PENALTY = f32(3)
export const FRONTLINE_WEIGHT = f32(15)

/* Pencarian jalur penuh untuk setiap provinsi kandidat akan berbiaya lebih
   besar daripada nilai keputusannya, jadi jaraknya dihitung sebagai jumlah
   lompatan yang dibatasi. */
const DISTANCE_CAP = 8

/*
 * Menentukan ke mana sebuah pasukan harus pergi setelah perang berjalan.
 *
 * Memutuskan untuk bertempur dan memutuskan di mana bertempur adalah dua
 * masalah terpisah: yang pertama diplomasi, ini operasi. Menggabungkannya
 * menghasilkan pasukan yang berbaris ke provinsi musuh mana pun yang kebetulan
 * pertama di larik.
 */
export class WarPlanner {
  constructor(
    private readonly world: WorldState,
    private readonly relations: Relations,
    private readonly land: ProvinceGraph,
  ) {}

  chooseTarget(army: Army, allArmies: ReadonlyMap<number, Army>): TargetChoice | null {
    const garrisons = new Map<number, number>()
    for (const other of allArmies.values()) {
      if (other.isDestroyed || other.nation === army.nation) continue
      garrisons.set(other.province, (garrisons.get(other.province) ?? 0) + other.count)
    }

    let best: TargetChoice | null = null
    let bestValue = Number.NEGATIVE_INFINITY

    for (let province = 0; province < this.world.provinces.count; province++) {
      const owner = this.world.provinces.controller[province]!
      if (owner === NO_OWNER || !this.relations.atWar(army.nation, owner)) continue

      const scored = this.#valueOf(province, army.province, garrisons)
      if (scored.value > bestValue) {
        bestValue = scored.value
        best = { province, value: scored.value, reason: scored.reason }
      }
    }

    return best
  }

  #valueOf(
    province: number,
    from: number,
    garrisons: ReadonlyMap<number, number>,
  ): { value: number; reason: string } {
    let value = f32(10)
    let reason = 'enemy ground'

    if (this.world.provinces.isCity[province] !== 0) {
      value = addF32(
        value,
        addF32(CITY_WEIGHT, mulF32(this.world.provinces.population[province]!, f32(4))),
      )
      reason = 'enemy city'
    }

    const defenders = garrisons.get(province) ?? 0
    if (defenders === 0) {
      value = addF32(value, UNDEFENDED_WEIGHT)
      if (reason === 'enemy ground') reason = 'undefended'
    } else {
      value = subF32(value, mulF32(defenders, f32(6)))
    }

    /* Provinsi yang berbatasan dengan tanah yang sudah dikuasai lebih berharga
       daripada hadiah jauh di dalam wilayah musuh: front yang maju bersama bisa
       dipasok, ujung tombak sendirian tidak. */
    if (this.#bordersOwnTerritory(province, this.world.provinces.controller[from]!)) {
      value = addF32(value, FRONTLINE_WEIGHT)
    }

    return {
      value: subF32(value, mulF32(this.#approximateDistance(from, province), DISTANCE_PENALTY)),
      reason,
    }
  }

  #bordersOwnTerritory(province: number, nation: number): boolean {
    for (const neighbour of this.land.neighboursOf(province)) {
      if (this.world.provinces.controller[neighbour] === nation) return true
    }

    return false
  }

  /* Pencarian melebar dengan jumlah lompatan, dibatasi. Antreannya larik dengan
     penunjuk kepala, bukan shift(): shift memindahkan seluruh isi larik setiap
     kali, dan pencarian ini dijalankan sekali per provinsi kandidat. */
  #approximateDistance(from: number, to: number): number {
    const seen = new Set<number>([from])
    const provinces: number[] = [from]
    const depths: number[] = [0]

    for (let head = 0; head < provinces.length; head++) {
      const province = provinces[head]!
      const depth = depths[head]!

      if (province === to) return depth
      if (depth >= DISTANCE_CAP) continue

      for (const neighbour of this.land.neighboursOf(province)) {
        if (seen.has(neighbour)) continue

        seen.add(neighbour)
        provinces.push(neighbour)
        depths.push(depth + 1)
      }
    }

    return DISTANCE_CAP
  }
}
